src/families.h

```diff
@@ -18,139 +18,106 @@
 #define SGDNET_FAMILIES_
 
 #include <memory>
+#include <RcppEigen.h>
 #include "math.h"
 #include "utils.h"
-
 namespace sgdnet {
 
 class Family {
 public:
-  Family(const std::vector<double>& y,
-         const unsigned n_samples,
-         const unsigned n_classes,
-         const double L_scaling)
-         : y(y),
-           n_samples(n_samples),
-           n_classes(n_classes),
-           L_scaling(L_scaling) {
-    // Initialize scaling and centering parameters for the response
-    y_center.resize(n_classes, 0.0);
-    y_scale.resize(n_classes, 1.0);
-    y_mat.setZero(n_samples, n_classes);
-    y_mat_scale.resize(n_classes, 1.0);
-  };
-
-  virtual double Link(const unsigned i) = 0;
-
-  virtual double Loss(const double prediction, const unsigned i) = 0;
-
-  virtual double Loss(const std::vector<double>& prediction,
-                      const unsigned i) = 0;
-
-  virtual void Gradient(std::vector<double>&       g_i,
-                        const std::vector<double>& prediction,
-                        const unsigned             i) = 0;
-
-  virtual void PreprocessResponse() = 0;
-
-  virtual double NullDeviance(const bool fit_intercept) = 0;
-
-  std::vector<double> StepSize(const double               max_squared_sum,
-                               const std::vector<double>& alpha_scaled,
-                               const bool                 fit_intercept) {
-    // Lipschitz constant approximation
-    std::vector<double> step_sizes;
-    step_sizes.reserve(alpha_scaled.size());
-
-    for (auto alpha_i : alpha_scaled) {
-      double L =
-        L_scaling*(max_squared_sum + static_cast<double>(fit_intercept))
-        + alpha_i;
-      double mu_n = 2.0*n_samples*alpha_i;
-      step_sizes.push_back(1.0 / (2.0*L + std::min(L, mu_n)));
-    }
-    return step_sizes;
-  }
+  Family(const double L_scaling) : L_scaling(L_scaling) {};
+
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) noexcept;
+
+  double Loss(const std::vector<double>& prediction,
+              const std::vector<double>& y,
+              const unsigned             i) const noexcept;
+
+  void Gradient(const std::vector<double>& prediction,
+                const std::vector<double>& y,
+                const unsigned             i,
+                std::vector<double>&       gradient) const noexcept;
+
+  double NullDeviance(const std::vector<double>& y,
+                      const bool                 fit_intercept,
+                      const unsigned             n_classes) const noexcept;
+
+  Eigen::MatrixXd
+  LambdaResponse(const std::vector<double>& y,
```

Comment:
What does `LambdaResponse` represent here? The name isn't clear to me.

Reply:
You're right. I was struggling to come up with something useful. This is basically a temporary, standardized version of the response that is used for computing LambdaMax. I'll see if I can come up with something better.